server/src/instant/reactive/session.clj

```diff
@@ -722,44 +662,69 @@
 ;; ------
 ;; System
 
-(defn group-fn [{:keys [item] :as _input}]
-  (let [{:keys [op session-id room-id]} item]
-    (case op
-      :transact
-      [:transact session-id]
+(defn group-key [{:keys [op session-id room-id q]}]
+  (case op
+    :transact
+    [:transact session-id]
+
+    (:join-room :leave-room :set-presence :client-broadcast :server-broadcast)
+    [:room session-id room-id]
+
+    (:add-query :remove-query)
+    [:query session-id q]
 
-      (:join-room :leave-room :set-presence :client-broadcast :server-broadcast)
-      [:room session-id room-id]
+    :refresh
+    [:refresh session-id]
 
-      (:add-query :remove-query)
-      (let [{:keys [q]} item]
-        [:query session-id q])
+    :refresh-presence
+    [:refresh-presence session-id room-id]
 
-      :refresh
-      [:refresh session-id]
+    :error
+    [:error session-id]
 
-      :refresh-presence
-      [:refresh-presence session-id room-id]
+    nil))
 
-      :error
-      [:error session-id]
+(defmulti combine
+  (fn [event1 event2]
+    [(:op event1) (:op event2)]))
 
-      nil)))
+(defmethod combine :default [_ _]
+  nil)
 
-(comment
-  (group-fn {:item {:session-id 1 :op :transact}})
-  (group-fn {:item {:session-id 1 :op :leave-room}})
-  (group-fn {:item {:session-id 1 :op :add-query :q {:users {}}}}))
+(defmethod combine [:refresh :refresh] [event1 event2]
+  (e2e-tracer/invalidator-tracking-step!
+   {:name          "skipped-refresh"
+    :tx-id         (:tx-id event1)
+    :tx-created-at (:tx-created-at event1)})
+  event2)
+
+(defmethod combine [:refresh-presence :refresh-presence] [event1 event2]
+  (update event2 :edits #(concat (:edits event1) %)))
```

Comment:
nit: should we use `into`? (I remember vaguely if we apply concat repeatedly on an item it can cause perf issues, but not 100% sure if it applies for this case

Reply:
Probably won’t matter here, but sure, lazy sequences are a loaded gun